### DISCUSSION TITLE
Unhandled Runtime Error Error: ctor is not a function

### DISCUSSION THREAD

**Discussion by danechitoaie** [vercel/next.js#37850] [Help]
I'm trying to use Suspense as https://nextjs.org/docs/advanced-features/dynamic-import
But I get the following errors when running `npm run dev`:

```
Unhandled Runtime Error
Error: ctor is not a function

Call Stack
updateDehydratedSuspenseComponent
node_modules/react-dom/cjs/react-dom.development.js (20662:0)
updateSuspenseComponent
node_modules/react-dom/cjs/react-dom.development.js (20362:0)
beginWork
node_modules/react-dom/cjs/react-dom.development.js (21624:0)
beginWork$1
node_modules/react-dom/cjs/react-dom.development.js (27426:0)
performUnitOfWork
node_modules/react-dom/cjs/react-dom.development.js (26557:0)
workLoopSync
node_modules/react-dom/cjs/react-dom.development.js (26466:0)
renderRootSync
node_modules/react-dom/cjs/react-dom.development.js (26434:0)
performConcurrentWorkOnRoot
node_modules/react-dom/cjs/react-dom.development.js (25738:0)
workLoop
node_modules/scheduler/cjs/scheduler.development.js (266:0)
flushWork
node_modules/scheduler/cjs/scheduler.development.js (239:0)
MessagePort.performWorkUntilDeadline
node_modules/scheduler/cjs/scheduler.development.js (533:0)
```

Any idea what's going on?

**Comment by icyJoseph** [comment]
Got a repo to take a look at? A snippet of the failing code? Or perhaps run `npx next info` - at least that'll give us some clue.

```tsx
import dynamic from 'next/dynamic'
import { Suspense } from 'react'

const DynamicHeader = dynamic(() => import('../components/header'), {
  suspense: true,
})

export default function Home() {
  return (
    <Suspense fallback={`Loading...`}>
      <DynamicHeader />
    </Suspense>
  )
}
```

This works, fine, I tried a few ways to break it and didn't get to reproduce your error. Do you have the right package versions installed? Again, a repo, or reproduction demo, snippet, would help a lot.

**Comment by danechitoaie** [comment]
@icyJoseph Here's a sample repo https://github.com/danechitoaie/suspense

```
git clone git@github.com:danechitoaie/suspense.git
cd suspense
npm install --legacy-peer-deps
npm run dev
```

Gives me `Error: ctor is not a function`


```
    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64
    Binaries:
      Node: 16.15.1
      npm: 8.12.2
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 12.1.6
      react: 18.2.0
      react-dom: 18.2.0

```

**Comment by icyJoseph** [comment]
I am not sure about this combination:

```tsx
const Challenge = dynamic(() => import("../components/Challenge"), {
  suspense: true,
  ssr: false,
});
```

As far as I understand the new Suspense, React SSR lets Suspenseful components resolve, but because we are telling it to not load the component during SSR, nothing is available as result of suspense, so the server errors outs.

**Comment by danechitoaie** [comment]
Now I'm getting 
```
Unhandled Runtime Error
Error: Only absolute URLs are supported
```

Isn't this how useSwr should be used?

```
export default function Challenge() {
    const { data } = useSWR(
        "/api/data",
        (...args) => fetch(...args).then((res) => res.json()),
        { suspense: true }
    );

    return (
        <>
           ...
        </>
    );
}

```

**Reply by icyJoseph** [reply]
Yeah, that's a really funky one, but at build time, and on the server, you cannot really reach out for API routes... In fact, what's going on here, is that `node-fetch` is kicking in, the fetch polyfill for Node.js.

This is where you gotta accept that Suspense for data-fetching, is not really there yet, https://github.com/vercel/swr/issues/1898 - Maybe if you do a window check on the fetcher and return null data if `typeof window === 'undefined'` - I am not sure why people rushed to try Suspenseful data-fetching day 1 of React 18, it was always clear that this feature would take a while longer.

**Comment by danechitoaie** [comment]
ok, makes sense.
I went back to `loading` property in favor of Suspense and works ok now

```
const Challenge = dynamic(() => import("../components/Challenge"), {
    ssr: false,
    loading: () => <div>loading...</div>,
});
```
